### ISSUE TITLE
AOT provideStore with initialState problem

### ISSUE THREAD

**Issue by loixlab** [ngrx/store#348]
Hi,

I am trying to provide the store with a localStorage saved state in AOT and can't find a solution...

This is how I have it so far:

```js
@NgModule({
    declarations: [
        AppComponent,
    ],
    imports: [
        CoreModule,
        SharedModule,
        AppRoutingModule,
        StoreModule.provideStore(rootReducer, LocalstorageService.loadState()),
    ],
    providers: [
        ENV_PROVIDERS,
    ],
    bootstrap: [AppComponent],
})
export class AppModule { }
```

The AOT build fails because it does not allow a function call in the metaData, So I added the possibility to execute a function inside the `StoreModule.provideStore` method in **ng2.js** like so

```js
StoreModule.provideStore = function (_reducer, _initialState) {
        if (typeof _initialState === 'function') {
            _initialState = _initialState();
        }
        return {
            ngModule: StoreModule,
            providers: provideStore(_reducer, _initialState)
        };
    };
```

I then changed the import to   
`StoreModule.provideStore(rootReducer, LocalstorageService.loadState)`

This **does work** in development server... but when I compile with AOT (without errors now) and run the files in the server, the state is not loaded from localStorage anymore...

Any help on this would be greatly appreciated! :)

**Comment by karolmie1** [comment]
Maybe you need to implement it differently. I just, when data is ready, dispatch new event containing new state, and i populate store with this state. Initial, totally empty initial state is still required. If that's the approach you would like to take I can link you to my repo, so you can see how exactly I did it. It works in AOT.

**Comment by loixlab** [comment]
Hi, thanks for your info.
I am following the official app and what the release 2.1.1 says by using provideStore as follow  ```StoreModule.provideStore(reducer, initialState)```

Why do you say that it is deprecated?

I am interested to see your solution altough it might be more a hack to make it work than a solution from @ngrx/store. The **initialState** parameter is meant to be used, and reading the localStorage (executing a function for that) to provide a state should be something quiet straightforward (and quiet a common use case for a SPA).  Like I said, I managed to make it work by tweeking their `Store.privideStore" method adding

```
if (typeof _initialState === 'function') {
    _initialState = _initialState();
}
````

If you could point me to your repo that'd be great, thanks! :)

**Comment by karolmie1** [comment]
@sebelga Wops, I got it mixed up (in other repo I use it was other way around) - I've removed the false info from my comment.

I've just used hmr reload setup:

```
export function reducer(state: any, action: any) {
    state = globalReducer(state, action);
    if ('production' === ENV) {
        let productionReducer: ActionReducer<IState> = combineReducers(reducers);
        return productionReducer(state, action);
    } else {
        let developmentReducer: ActionReducer<IState> = compose(storeFreeze, combineReducers)(reducers);  // tslint:disable-line:max-line-length
        return developmentReducer(state, action);
    }
}

function globalReducer(state: IState, action: any) {
    switch (action.type) {
        case AppActions.ActionTypes.SET_APP_STATE: {
            let payload = (action as AppActions.SetAppStateAction).payload;
            return payload;
        }
        default: {
            return state;
        }
    }
}
```
In file: https://bitbucket.org/kmierzej/fragaria-webapp/src/0cf15ce1b85854d64ecd22b3701f9be08128280b/src/app/store/app.reducers.ts?at=master&fileviewer=file-view-default

**Comment by loixlab** [comment]
Oh ok I see.. thanks. I will keep this as an alternative option. I would still like to see @ngrx/store accept a function for the initialState :)
thanks!

**Comment by jcroll** [comment]
I am running into this issue as well. I am using a factory to create my initial state and root reducer and the angular cli is throwing errors because of static analysis:

```typescript
export function createStoreFactory(): StoreFactory {
  let storeFactory =  new StoreFactory();

  return storeFactory
  .addModel(Account)
  .addModel(User);
}

export const storeFactory = createStoreFactory();

export const initialState = storeFactory.createInitialState();

export const rootReducer = storeFactory.createRootReducer();

```
...
```typescript

@NgModule({
  imports: [
    StoreModule.provideStore(
      rootReducer,
      initialState
    ),
```
...
```
frontend_1  | Hash: 0723a1c291c6fb8c9493                                                                 
frontend_1  | Time: 36100ms
frontend_1  | chunk    {0} main.bundle.js, main.bundle.js.map (main) 201 kB {2} [initial] [rendered]
frontend_1  | chunk    {1} styles.bundle.js, styles.bundle.js.map (styles) 363 kB {3} [initial] [rendered]
frontend_1  | chunk    {2} vendor.bundle.js, vendor.bundle.js.map (vendor) 4.6 MB [initial] [rendered]
frontend_1  | chunk    {3} inline.bundle.js, inline.bundle.js.map (inline) 0 bytes [entry] [rendered]
frontend_1  | 
frontend_1  | ERROR in Error encountered resolving symbol values statically. Calling function 'createStoreFactory', function calls are not supported. Consider replacing the function or lambda with a reference to an exported function, resolving symbol storeFactory in /usr/share/frontend/src/app/stores/create-store-factory.ts, resolving symbol rootReducer in /usr/share/frontend/src/app/stores/create-store-factory.ts, resolving symbol rootReducer in /usr/share/frontend/src/app/stores/create-store-factory.ts, resolving symbol AppModule in /usr/share/frontend/src/app/app.module.ts, resolving symbol AppModule in /usr/share/frontend/src/app/app.module.ts, resolving symbol AppModule in /usr/share/frontend/src/app/app.module.ts
```

I currently see no way to accomadate this other than:
```
if (typeof _initialState === 'function') {
    _initialState = _initialState();
}
```

as [sebelga said](https://github.com/ngrx/store/issues/348#issuecomment-285316117).

Related: https://github.com/ngrx/store/issues/190

**Comment by jcroll** [comment]
Check [this comment](https://github.com/ngrx/store/issues/154#issuecomment-227651036) for a way to potentially create the root reducer statically with a factory:

> In one of my projects I have created a store service which takes care of adding new reducers to an existing set of reducers. Each other service can now add its reducer by itself.

> Let me know if it helps, or if I can make something better:
https://gist.github.com/philipooo/d4acbcf314b69df0695eb05f55c03876

Basically use `Store.replaceReducer()`

Edit: got it to build using this method

**Comment by loixlab** [comment]
thanks for the info @jcroll . I might (will have to) do something similar.  It would have been nice to be able to do something similar that you can do in React (https://egghead.io/lessons/javascript-redux-persisting-the-state-to-the-local-storage) without having to replace reducers when we just want to set the initial store state.

Do you think that with the solution of replacing reducer after the app is booted, we can provide a "incomplete" state ? I mean, I don't save all my state in localStorage, just part of it, so I wonder if the Store.replaceReducer replaces the state or if it **merges** the initialState with the one provided

thanks.

**Comment by jcroll** [comment]
For the reducer:
```
let intitalState = localStorage.getItem('initialState');

return function reduce(state = initalState, action: Action) {

**Comment by loixlab** [comment]
Ok thanks!

**Comment by loixlab** [comment]
If anyone bumps into this, I finally got the solution inspired by the "ngrx-store-localstorage" package.

```js
// Your reducers
export const reducers: any = {
    app: AppReducer,
    user: UserReducer,
    ....
};

// add a middleware for the compose
const localStorageSync = (reducer: any) => {
    return function (state, action: any) {
        if (action.type === '@ngrx/store/init') {
            // only for the init action we update the state. Here we can call our function :)
            state = Object.assign({}, state, StoreService.loadState());
        }

        const nextState = reducer(state, action);
        return nextState;
    }
}

// rootReducer is what you provide to StoreModule.provideStore(rootReducer)
export const rootReducer: ActionReducer<State> = compose(localStorageSync,  combineReducers)(reducers);
```

```js
// this is the StoreService.loadState (static)

static loadState() {
    try {
        const serializedState = localStorage.getItem('state');

        if (serializedState === null) {
            return undefined;
        }

        return JSON.parse(serializedState);
    } catch (err) {
        return undefined;
    }
}
```

**Comment by kylecordes** [comment]
@sebelga  Although probably outside the scope of this item, I am curious what solution you have used for the problem of the "shape" of the state data changing over time.  By this I mean, version N of the software expects a certain set of fields in the state, version N+1 expects a slightly different set of fields in the state.  I have used code similar to what you have written above, and it seems to open up problematic failure modes - in which once the software is broken, it remains broken even when the user refreshes the page.  At worse case, it can result in having to tell users that they must make many clicks in their browser configuration to manually clear local storage for the application.

**Comment by orizens** [comment]
hi @sebelga - see my first comment on #385

**Comment by loixlab** [comment]
@kylecordes that's a good point but I haven't had any problem so far as the state is merged.

[EDIT]: I was wrong sorry, effectively if the shape of the slice of the store I saved to localStorage has less properties than the new state (from an app version update), those properties are "lost" and do not appear in the store.

**Comment by mustafaekim** [comment]
I have a similar problem and I didnt understand how I can overcome this problem. I try to init my state with:

```ts
export function initAppState(): AppState {
    return {
        userInteraction: new UserInteraction()
        , authentication: new AsyncContainer<LoggedInAccount>()
        , eventManager: new AsyncContainer<EventManager>()
        , myEvents: new AsyncContainer<Event[]>()
    }
}
```

and I am getting the error:

> Error encountered resolving symbol values statically. Calling function 'UserInteraction', function calls are not supported. Consider replacing the function or lambda with a reference to
>  an exported function, resolving symbol initAppState in

How else can I initiate the state ?

(that solution works for webpack but does not for AOT. I am looking for a solution for AOT)

**Comment by bfricka** [comment]
It's unfortunate that this thread has received zero response from the maintainers. Fixing this would be as simple as allowing a function to be passed as `initialState`. And in ng2.ts just look for:

```ts
export function _initialStateFactory(initialState, reducer) {
  if (!initialState) {
    return reducer(undefined, { type: Dispatcher.INIT });
  }

  return typeof initialState === 'function'
    ? initialState()
    : initialState;
}
```
If there's interest, I'll open a PR for this.

**Comment by bfricka** [comment]
One more thing: It's totally possible to implement this with a meta-reducer that sits on top and looks for this specific case. Here's how you'd implement this.

Meta-reducer:
```ts
import {Action, Dispatcher} from '@ngrx/store';

export const initialStateReducer = (_initialState: any) => (reducer: any) => {
    const initialState = typeof _initialState === 'function'
        ? _initialState()
        : _initialState;

    return (state: any, action: Action) => {
        if (action.type === Dispatcher.INIT) {
            state = initialState;
        }

        return reducer(state, action);
    };
};
```

Root reducer:
```ts
const initialStateObjectFromSomewhere = {users: ['bob', 'brenda']};

function getInitialState() {
    return {
        router: {path: location.pathname + location.search},
        ...initialStateObjectFromSomewhere,
    };
}

const reducer = compose(
    initialStateReducer(getInitialState),
    combineReducers,
)(reducers);

export function rootReducer(state: any, action: any) {
    return reducer(state, action);
}
```
In this case you could also just make your rootReducer do the initial state stuff. This is tested and working with AoT.

**Comment by rupeshtiwari** [comment]
@bfricka Thanks I have one  question if I want to do state restore where initial state comes from database. Then is there any way to get the initial state from some injectable service. Since this is reducer and reducer can not get any injectable services. Do we have to do it in effects and then dispatch some action which this metareducer should listen and set application state ?

**Comment by rupeshtiwari** [comment]
@bfricka Here is my code which is working
 
```
@Injectable()
export class SaveEffects {
  @Effect()
  save$: Observable<Action> = this.actions$
    .ofType(fromSave.SAVE)
    .map(toPayload)
    .switchMap(state => saveAppState(state))
    .map((o) => fromSave.saveComplete(o))
    .catch((e) => of(fromSave.saveFail(e)));

  @Effect()
  restore$: Observable<Action> = this.actions$
    .ofType(Dispatcher.INIT)
    .switchMap(() => this.localStorageService.getAppState())
    .map((s) => appActions.appStateAvailable(s));
  constructor(
    private actions$: Actions, private localStorageService: LocalStorageService
  ) { }
}

index.ts //root reducer 
export const restoreStateReducer = (reducer: any) => {
    return (state: any, action: Action) => {
        if (action.type === appActions.APP_STATE_AVAILABLE) {
            state = action.payload;
        }
        return reducer(state, action);
    };
};


export const developmentReducer = compose(
    restoreStateReducer,
    storeFreeze,
    storeLogger(),
    combineReducers)(reducers);
```

**Comment by bfricka** [comment]
Since you're composing the `restoreStateReducer`, I don't see why this wouldn't work? The main thing you'd need to make to AoT compatible is for your main reducer to be a function rather than an expression. So:

```ts
// index.ts
export function rootReducer(state: any, action: any) {
  return developmentReducer(state, action);
}
// app module
@NgModule({
  imports: [StoreModule.provideStore(rootReducer)]
})
```

That said, I don't use `@Effects` so maybe I'm missing something. If you're just rehydrating state using local storage, and you want that to populate the initial state, you're probably better off looking at something like [ngrx-store-localstorage](https://github.com/btroncone/ngrx-store-localstorage) for inspiration.

The key thing that being able to pass a function as `initialState` allows is for the population of initial state to be decoupled from the reducer chain while allowing dynamic state. Otherwise, you can intercept it and populate it as described previously.

As far as populating initial state w/ deps, the only way I can think is to override the provider (ng2.ts). E.g.

```ts
// app module
import {INITIAL_REDUCER, INITIAL_STATE, Dispatcher} from '@ngrx/store';
// This takes the place of the second param to `providerStore`
const staticInitialState = {foo: {}};

export function getInitialState(localStorageService: LocalStorageService, reducer: any) {
  const initialState = {...staticInitialState, ...localStorageService.getAppState()};
  return reducer(initialState, Dispatcher.INIT);
}

@NgModule({
  providers: [
    {provide: INITIAL_STATE, useFactory: getInitialState, deps: [LocalStorageService, INITIAL_REDUCER]}
  ]
})
```

**Comment by robwormald** [comment]
support for an initial state function just landed in v4.

**Comment by slawojstanislawski** [comment]
It landed here: https://github.com/ngrx/platform/commit/1a166ec3f320c23fd320e23d844cc3493f4e2a93 -
it's tagged 4.0.5, can someone please publish 4.0.5 to npm? The last published version is 4.0.3. Thank you in advance.

**Comment by vfabregat** [comment]
@slawojstanislawski the platform version is the 4.0.5. but the store version that included the fix is the 4.0.3
